### PR TITLE
[NonPoolSuperfluid] Use full routes instead of only one pool for pricing

### PR DESCRIPTION
**Pull request by nicolaslara**
Closes: #XXX

## What is the purpose of the change

Allow native superfluid to use a full route instead of only one pool

## Testing and Verifying
tests updated
## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented?
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/poolmanager/types/routes.go [hunk 1]
```diff
@@ -167,3 +167,7 @@ func ValidateSwapAmountOutSplitRoute(splitRoutes []SwapAmountOutSplitRoute) erro
 
 	return nil
 }
+
+func (route *SwapAmountInRoute) Equal(routeB *SwapAmountInRoute) bool {
+	return route.PoolId == routeB.PoolId && route.TokenOutDenom == routeB.TokenOutDenom
+}
```
[the remark this answers is not in the document]
Comment (nicolaslara) [reply x/poolmanager/types/routes.go:173]: this is needed to comply with the interface

File: x/superfluid/keeper/integration_test.go [hunk 2]
```diff
@@ -470,13 +480,22 @@ func (s *TestSuite) TestNativeSuperfluid() {
 		},
 		validatorRewards)
 	s.Require().NoError(err)
-	s.Require().Equal(1, len(validatorRewards.Rewards.Rewards))
-	// After unstaking we will check that those rewards were properly distributed to the delegator
+	// Validators get stake and uosmo. I think this is an issue with test setup
+	s.Require().Equal(2, len(validatorRewards.Rewards.Rewards))
```
Comment (czarcas7ic) [inline x/superfluid/keeper/integration_test.go:484]: Would like to get to the bottom of this prior to merge

File: x/superfluid/keeper/epoch.go [hunk 3]
```diff
@@ -161,15 +159,13 @@ func (k Keeper) UpdateOsmoEquivalentMultipliers(ctx sdk.Context, asset types.Sup
...
-
 		// Twap price is calculated from the last 5 minutes
 		startTime := ctx.BlockTime().Add(-5 * time.Minute) // TODO: make this configurable
-		price, err := k.twapk.GetArithmeticTwapToNow(ctx, asset.PricePoolId, bondDenom, asset.Denom, startTime)
+		price, err := k.twapk.GetMultiPoolArithmeticTwapToNow(ctx, asset.PriceRoute, asset.Denom, bondDenom, startTime)
 		if err != nil {
 			return sdkerrors.Wrap(err, "failed to get twap price")
 		}
-		k.SetOsmoEquivalentMultiplier(ctx, newEpochNumber, asset.Denom, osmomath.NewDec(1).Quo(price))
+		k.SetOsmoEquivalentMultiplier(ctx, newEpochNumber, asset.Denom, price)
```
[the remark this answers is not in the document]
Comment (nicolaslara) [reply x/superfluid/keeper/epoch.go:168]: before we were doing:
` k.twapk.GetArithmeticTwapToNow(ctx, asset.PricePoolId, bondDenom, asset.Denom, startTime)`
and now
 `k.twapk.GetMultiPoolArithmeticTwapToNow(ctx, asset.PriceRoute, asset.Denom, bondDenom, startTime)`
 Notice how the denoms used for asset and quote changed. So no need to invert it if we get the right price directly

File: x/superfluid/keeper/integration_test.go (hunk 2 above)
Comment (nicolaslara) [reply x/superfluid/keeper/integration_test.go:484]: The reason for this is that what gets distributed is every collected fee. That comes from checking the fee collector's balances here: https://github.com/cosmos/cosmos-sdk/blob/main/x/distribution/keeper/allocation.go#L23

Somewhere along the line, someone is sending stake to the fee collector (haven't figured out where, but I think that's ok). The behaviour of distributing that fee is correct though

**Review by czarcas7ic — APPROVED** [review]
Would like to see the GetMultiPoolArithmeticTwapToNow test in api_test.go, but other than that it looks fine. I will approve but I personally would like to review the test if possible before merge.

File: x/twap/api.go [hunk 4]
```diff
@@ -141,3 +146,40 @@ func (k Keeper) getTwapToNow(
...
 	return k.getMostRecentRecord(ctx, poolId, asset0Denom, asset1Denom)
 }
+
+// GetMultiPoolArithmeticTwapToNow returns the price of two assets through multiple pools.
+// The price is calculated by taking the arithmetic TWAP of the two assets in each pool and multiplying
+// them together.
+// Only pools with two assets are considered.
+// For each pool n, its base asset will be the quote asset of pool n-1. The first pool's base asset is specified
+// in baseAssetDenom and the last pool's quote asset is specified in quoteAssetDenom.
+func (k Keeper) GetMultiPoolArithmeticTwapToNow(
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/twap/api.go:156]: We need some warnings here for this not actually being the correct TWAP of the pair.

We want the average value of $price_{pool1} * price_{pool2}$ which is $$\text{scaling factor} * \sum_{b \in blocks} \frac{price_{pool1} * price_{pool2}}{ \text{b.block time}}$$

This PR is returning $$(\sum_{b \in blocks} \frac{price_{pool1}}{ b.block_time} * (\sum_{b \in blocks} \frac{price_{pool2}}{b.block\_time}$$

These are not the same.

So we need to be careful about the precise usages here, or consider bucketing into many smaller intervals for a big TWAP, to improve accuracy.

[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/twap/api.go:156]: TL;DR average(a*b) != average(a)*average(b)

We can add a comment saying this is likely an acceptable approximation error in the following two cases:
- if one asset is a stable pair we are fine assuming is basically constant for the time range
- If all assets involved are major assets, and we rely on a proto-rev cyclic arb assumption. Then the result should be within a (guess-timate) 1-3% error tolerance. (roughly ~.3% * num_pools in path to close arb)

[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/twap/api.go]: Explanation added here https://github.com/osmosis-labs/osmosis/pull/8276/commits/35097231ce4bd16db0efeef04b350a80199ffc8e, thanks for breaking this down

File: x/superfluid/keeper/stake_test.go [hunk 5]
```diff
@@ -158,6 +159,8 @@ func (s *KeeperTestSuite) TestSuperfluidDelegate() {
 
 func (s *KeeperTestSuite) TestValidateLockForSFDelegate() {
 	lockOwner := s.TestAccs[0]
+	params, _ := s.App.StakingKeeper.GetParams(s.Ctx)
```
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/superfluid/keeper/stake_test.go]: https://github.com/osmosis-labs/osmosis/pull/8276/commits/5ed237979719024a804e5a838dedd8d5e8ca229b

**Comment by czarcas7ic** [comment]
```
	baseAsset := baseAssetDenom
	quoteAsset := route[0].TokenOutDenom
	for _, pool := range route {
		twap, err := k.GetArithmeticTwapToNow(ctx, pool.PoolId, baseAsset, quoteAsset, startTime)
		if err != nil {
			return osmomath.Dec{}, errorsmod.Wrapf(err, "failed to get arithmetic twap for pool %d", pool.PoolId)
		}
		price = price.Mul(twap)
		// Update the assets to the next pool's base and quote assets
		baseAsset = quoteAsset
		quoteAsset = pool.TokenOutDenom
	}

	return price, nil
}
```

I'm pretty sure this is wrong, we shouldn't be setting the quoteAsset as pool.TokenOutDenom within the loop itself

**Comment by czarcas7ic** [comment]
```
	baseAsset := baseAssetDenom
	quoteAsset := route[0].TokenOutDenom
	for _, pool := range route {
		twap, err := k.GetArithmeticTwapToNow(ctx, pool.PoolId, baseAsset, quoteAsset, startTime)
		if err != nil {
			return osmomath.Dec{}, errorsmod.Wrapf(err, "failed to get arithmetic twap for pool %d", pool.PoolId)
		}
		price = price.Mul(twap)
		// Update the assets to the next pool's base and quote assets
		baseAsset = quoteAsset
		quoteAsset = pool.TokenOutDenom
	}

	return price, nil
}
```

I am pretty sure the way we set `quoteAsset = pool.TokenOutDenom` inside the loop is wrong, we want to set it to the next pool's tokenOut, not the pool we just finished. Otherwise we are route[0].TokenOutDenom twice as the quote

File: x/twap/api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/twap/api.go]: Also, tests and bug in the method corrected here https://github.com/osmosis-labs/osmosis/pull/8276/commits/be2a32b7f42454b79f0dcb103eb8a7a6d9286204

File: x/superfluid/keeper/integration_test.go (hunk 2 above)
Comment (czarcas7ic) [reply x/superfluid/keeper/integration_test.go]: Double emission in test fixed here https://github.com/osmosis-labs/osmosis/pull/8276/commits/63e48b5612f6c9b8bc542884dd1cb256f66cd106

File: x/twap/api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicolaslara) [reply x/twap/api.go]: I think solving it with documentation makes sense here, as what we want is the "current price of an asset through a route, based on each historical twap" even if that isn't the theoretical twap of the route.

@ValarDragon I understand that they aren't the same for a twap of the route, but do you agree that the product of averages is enough to determine an accepted price of the route at the time queried? Just double checking my understanding here, but I think for errors to accumulate we would need each pool in the route to be mispriced in the same direction.

**Review by PaddyMc — APPROVED** [review]
:muscle:

File: x/twap/api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/twap/api.go]: I think its enough to determine an acceptable price in the circumstances were concerned about! The general case requires us to be careful / think about no possible tail explosions. (E.g. we don't weight pools by liquidity at time, so if you averaged across routes, or even in a route with one low liquidity pool, you could get non-sensical values)

**Pull request merged**